src/libertem/io/dataset/base/backend_buffered.py

```diff
@@ -0,0 +1,262 @@
+            else:
+                min_per_file[fileno] = min(part_rr[1], min_per_file[fileno])
+
+            if fileno not in max_per_file:
+                max_per_file[fileno] = part_rr[2]
+            else:
+                max_per_file[fileno] = max(part_rr[2], max_per_file[fileno])
+
+            if fileno not in payload_bytes:
+                payload_bytes[fileno] = 0
+            payload_bytes[fileno] += part_rr[2] - part_rr[1]
+
+    min_fill_factor = -1
+    max_buffer_size = 0
+
+    for k in min_per_file:
+        read_size = max_per_file[k] - min_per_file[k]
+        fill_factor = payload_bytes[k] / (read_size)
+        # print(k, fill_factor, max_per_file[k], min_per_file[k], payload_bytes[k])
+        if fill_factor < min_fill_factor or min_fill_factor == -1:
+            min_fill_factor = fill_factor
+        if read_size > max_buffer_size:
+            max_buffer_size = read_size
+    return min_fill_factor, max_buffer_size, min_per_file, max_per_file
+
+
+class BufferedBackend(IOBackend, id_="buffered"):
+    def __init__(self, max_buffer_size=16*1024*1024, sparse_threshold=0.9):
+        """
+        I/O backend using a buffered reading strategy. Useful for slower
+        media like HDDs, where seeks cause performance drops.
+
+        This does not perform optimally on SSDs, there
+        :class:`~libertem.io.dataset.base.MMapBackend` should be used
+        instead.
+
+        Parameters
+        ----------
+        max_buffer_size : int
+            Maximum buffer size, in bytes
+
+        sparse_threshold : float
+            When to switch to a sparse reading strategy, as a fraction
+            bytes_used / bytes_read.
+        """
+        self._max_buffer_size = max_buffer_size
+        self._sparse_threshold = sparse_threshold
+
+    @classmethod
+    def from_json(cls, msg):
+        """
+        Construct an instance from the already-decoded `msg`.
+        """
+        raise NotImplementedError("TODO! implement me!")
+
+    def get_impl(self):
+        return BufferedBackendImpl(
+            max_buffer_size=self._max_buffer_size,
+            sparse_threshold=self._sparse_threshold,
+        )
+
+
+class BufferedBackendImpl(IOBackendImpl):
+    def __init__(self, max_buffer_size, sparse_threshold):
+        super().__init__()
+        self._max_buffer_size = max_buffer_size
+        self._sparse_threshold = sparse_threshold
+        self._buffer_pool = BufferPool()
+
+    def need_copy(
+        self, decoder, roi, native_dtype, read_dtype, tiling_scheme=None, fileset=None,
+        sync_offset=0, corrections=None,
+    ):
+        return True  # we always copy in this backend
+
+    def get_read_and_decode(self, decode):
+        key = (decode, "read")
+        if key in _r_n_d_cache:
+            return _r_n_d_cache[key]
+        r_n_d = _make_buffered_reader_and_decoder(decode=decode)
+        _r_n_d_cache[key] = r_n_d
+        return r_n_d
+
+    def get_max_io_size(self):
+        return 16*2**20
```

Comment:
Why this size? Is it related to the `max_buffer_size` parameter? I don't understand why this is fixed at 16 MB while the `max_buffer_size` is a variable user parameter, to be honest...

Reply:
Sorry for the confusion - the `max_buffer_size` parameter was related to selecting sparse/dense reading strategies. As it turns out, in practice the dense strategy doesn't have large overheads, so for now I've not implemented the sparse strategy.

I've now hooked up the `max_buffer_size` parameter to `get_max_io_size`, so there should no longer be a "magic number".